### PR TITLE
Policy Service Tests

### PR DESCRIPTION
**Pull request by justindbaur**
This is a part of continued work on #419. 

It adds tests for:
- `PolicyService`
- `SsoConfigService`

I also took this opportunity to do a small cleanup in `PolicyService`. I took advantage of calls that were already made so that they didn't have to be made again.

### CODE REVIEW

**Comment by justindbaur** [comment]
Hey @MGibson1,  I am working on some tests for `PolicyService.cs` but I have a few questions regarding it's current implementation.  On [line 99](https://github.com/bitwarden/server/blob/master/src/Core/Services/Implementations/PolicyService.cs#L99) and [line 110](https://github.com/bitwarden/server/blob/master/src/Core/Services/Implementations/PolicyService.cs#L110) you are calling `_orgranizationRepository.GetByIDAsync(policy.OrganizationId);` but that exact same call (as far as it looks to me) is made on [line 37](https://github.com/bitwarden/server/blob/master/src/Core/Services/Implementations/PolicyService.cs#L37) and is then ensured to not be null right after. Could we not just utilize the existing `org` variable that is still in scope instead of calling the database one more time?

Also you are storing `var now = DateTime.UtcNow;` on [line 75](https://github.com/bitwarden/server/blob/master/src/Core/Services/Implementations/PolicyService.cs#L75) and possibly using it to set the `CreationDate` but you are calling `DateTime.Now` again for the `RevisionDate` instead of reusing `now` even though it is already in scope. Could I make the change to use `now` on [line 129](https://github.com/bitwarden/server/blob/master/src/Core/Services/Implementations/PolicyService.cs#L129).

**Comment by MGibson1** [comment]
@justindbaur -- Both of those observations look true to me as well. I'm totally for cleaning up these kinds of things. One of the benefits of adding these tests is going over these classes in a holistic sense and clearing this stuff up. Tests exist to improve the prod code, after all.

Blame shows that likely both variables were just forgotten when implementing some features further down the method -- it happens, you know? 🤷

**Comment by justindbaur** [comment]
@MGibson1 Awesome, I think my fear was that some call in between the first one and the time it gets called a second time that one of the method calls in between makes a stored procedure call and it could have updated a value on that organization that might matter. My mocks of those calls wouldn't reflect that though so I could introduce an issue. But if you agree that it is a relic then I will make those fixes on this PR. Thanks!

**Review by MGibson1 — COMMENTED** [review]
Can you split out the coverage stuff into another PR?

People have strong opinions about dev tools and data points. I for one don't like coverage as a blocker for merging or really even as a tracked metric. To me, it's mostly useful in the testing feedback loop.

**Comment by justindbaur** [comment]
Yeah I can. I put it all into one commit so it would be easy to take out. Nothing I added in would be any sort of blocker, and in it's current state it wouldn't even be a tracked metric. It would only be something that someone could run and look at if they want. I have just found myself adding these things to the a few times now while I am testing to make sure I am covering the branches. Up until now I have just been not committing any of it. I will pull it out though and make it's own PR so it can be challenged on it's own.

**Comment by MGibson1** [comment]
Thanks, I didn't mean that I'm opposed to what you have, just that we should discuss separately. Thanks for the push -- enabling code coverage is a boon for sure

File: test/Core.Test/Services/PolicyServiceTests.cs [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+                .Returns(Task.FromResult(new Core.Models.Table.Policy { Enabled = true }));
+
+            var badRequestException = await Assert.ThrowsAsync<BadRequestException>(
+                () => sutProvider.Sut.SaveAsync(policy,
+                    Substitute.For<IUserService>(),
+                    Substitute.For<IOrganizationService>(),
+                    Guid.NewGuid()));
+
+            Assert.Contains("Single Sign-On Authentication policy is enabled.", badRequestException.Message, StringComparison.OrdinalIgnoreCase);
+        }
```
Comment (MGibson1) [inline test/Core.Test/Services/PolicyServiceTests.cs:73]: Please assert no calls to `UpsertAsync` or `LogPolicyEventAsync`

File: test/Core.Test/Services/PolicyServiceTests.cs [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+                .Returns(Task.FromResult(new Core.Models.Table.Policy { Enabled = false }));
+
+            var badRequestException = await Assert.ThrowsAsync<BadRequestException>(
+                () => sutProvider.Sut.SaveAsync(policy,
+                    Substitute.For<IUserService>(),
+                    Substitute.For<IOrganizationService>(),
+                    Guid.NewGuid()));
+
+            Assert.Contains("Single Organization policy not enabled.", badRequestException.Message, StringComparison.OrdinalIgnoreCase);
+        }
```
Comment (MGibson1) [inline test/Core.Test/Services/PolicyServiceTests.cs:97]: Please assert no calls to `UpsertAsync` or `LogPolicyEventAsync`

File: test/Core.Test/Services/PolicyServiceTests.cs [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+            SetupOrg(sutProvider, policy.OrganizationId, null);
+
+            var badRequestException = await Assert.ThrowsAsync<BadRequestException>(
+                () => sutProvider.Sut.SaveAsync(policy,
+                    Substitute.For<IUserService>(),
+                    Substitute.For<IOrganizationService>(),
+                    Guid.NewGuid()));
+
+            Assert.Contains("Organization not found", badRequestException.Message, StringComparison.OrdinalIgnoreCase);
+        }
```
[the remark this answers is not in the document]
Comment (justindbaur) [reply test/Core.Test/Services/PolicyServiceTests.cs:38]: Absolutely! Just fixed this one and the others.

**Review by MGibson1 — APPROVED** [review]
Thank you for these tests!

This is a sensitive area of the codebase we need to get right, adding these is a huge help as we expand out our policies offering. Thank you again

**Pull request merged**